include/exadg/incompressible_navier_stokes/spatial_discretization/operator_consistent_splitting.cpp

```diff
@@ -0,0 +1,776 @@
+      }
+      else
+      {
+        AssertThrow(false,
+                    dealii::ExcMessage("Boundary type of face is invalid or not implemented."));
+      }
+    }
+    pressure.integrate_scatter(dealii::EvaluationFlags::values, dst);
+  }
+}
+
+template<int dim, typename Number>
+void
+OperatorConsistentSplitting<dim, Number>::rhs_ppe_laplace_add(VectorType &   dst,
+                                                        double const & evaluation_time) const
+{
+  ProjectionBase::do_rhs_ppe_laplace_add(dst, evaluation_time);
+}
+
+template<int dim, typename Number>
+unsigned int
+OperatorConsistentSplitting<dim, Number>::solve_pressure(VectorType &       dst,
+                                                   VectorType const & src,
+                                                   bool const         update_preconditioner) const
+{
+  return ProjectionBase::do_solve_pressure(dst, src, update_preconditioner);
+}
+
+template<int dim, typename Number>
+void
+OperatorConsistentSplitting<dim, Number>::interpolate_velocity_dirichlet_bc(VectorType &   dst,
+                                                                      double const & time) const
+{
+  this->evaluation_time = time;
+
+  dst = 0.0;
+
+  VectorType src_dummy;
+  this->get_matrix_free().loop(&This::cell_loop_empty,
+                               &This::face_loop_empty,
+                               &This::local_interpolate_velocity_dirichlet_bc_boundary_face,
+                               this,
+                               dst,
+                               src_dummy);
+}
+
+template<int dim, typename Number>
+void
+OperatorConsistentSplitting<dim, Number>::local_interpolate_velocity_dirichlet_bc_boundary_face(
+  dealii::MatrixFree<dim, Number> const & matrix_free,
+  VectorType &                            dst,
+  VectorType const &,
+  Range const & face_range) const
+{
+  unsigned int const dof_index  = this->get_dof_index_velocity();
+  unsigned int const quad_index = this->get_quad_index_velocity_nodal_points();
+
+  FaceIntegratorU integrator(matrix_free, true, dof_index, quad_index);
+
+  for(unsigned int face = face_range.first; face < face_range.second; face++)
+  {
+    dealii::types::boundary_id const boundary_id = matrix_free.get_boundary_id(face);
+
+    BoundaryTypeU const boundary_type =
+      this->boundary_descriptor->velocity->get_boundary_type(boundary_id);
+
+    if(boundary_type == BoundaryTypeU::Dirichlet or boundary_type == BoundaryTypeU::DirichletCached)
+    {
+      integrator.reinit(face);
+      integrator.read_dof_values(dst);
+
+      for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+      {
+        unsigned int const local_face_number = matrix_free.get_face_info(face).interior_face_no;
+
+        unsigned int const index = matrix_free.get_shape_info(dof_index, quad_index)
+                                     .face_to_cell_index_nodal[local_face_number][q];
```

Comment:
this would not work for H(div) would it?
I grepped `face_to_cell_index_nodal` and see it is used in some places in the library.
Probably that was a discussion point before ...
I guess since it is similar to the dual splitting scheme we go with this until we port the strategy from Martin's branch over to ExaDG.

Reply:
thanks for the comment, lets leave it for now, if needed we change it later for the HODS and this scheme at once.